replica/table.cc

```diff
@@ -1915,11 +1916,25 @@ compaction_group::update_sstable_sets_on_compaction_completion(sstables::compact
             for (auto& [cg, d] : _cg_desc) {
                 d.new_main_sstables = co_await _builder.build_new_list(*cg->main_sstables(), _t._compaction_strategy.make_sstable_set(_t._schema),
                                                                   d.desc.new_sstables, d.desc.old_sstables);
+
+                if (!d.desc.old_sstables.empty() &&
+                        (d.new_main_sstables->size() > cg->main_sstables()->size() + d.desc.new_sstables.size() - d.desc.old_sstables.size())) {
+                    // The new_main_sstables size is larger than expected, which implies that the old_sstables
+                    // were not removed from the set. This can happen if the input sstables were picked up from
+                    // the maintenance set during an offstrategy or scrub compaction. So, remove the old sstables
+                    // from the maintenance set. No need to add the new sstables to the maintenance set though,
+                    // as they are always added to the main set.
+                    d.new_maintenance_sstables = co_await _builder.build_new_list(*_cg.maintenance_sstables(), std::move(*_t.make_maintenance_sstable_set()), {}, d.desc.old_sstables);
+                }
             }
         }
         virtual void execute() override {
             for (auto&& [cg, d] : _cg_desc) {
                 cg->set_main_sstables(std::move(d.new_main_sstables));
+                if (d.new_maintenance_sstables) {
+                    // offstrategy or scrub compaction - replace the maintenance set
+                    cg->set_maintenance_sstables(std::move(d.new_maintenance_sstables.value()));
+                }
             }
             // FIXME: the following is not exception safe
             _t.refresh_compound_sstable_set();
```

Comment:
only old sstables from main should be removed from the backlog tracker below. maybe build_new_list() interface can be adjusted to return which sstables were actually removed from the set (using pair or struct), and then you can use that info to simply condition to trigger maintenance set rebuild, and adjust backlog tracker appropriately.

Reply:
Done.